File: bench/go/arena.go
package main

import "fmt"

type Node struct {
	val  int32
	next *Node
}

func main() {
	const N int32 = 5_000_000

	nodes := make([]Node, N)
	var head *Node = nil
	for i := int32(0); i < N; i++ {
		n := &nodes[i]
		n.val = i
		n.next = head
		head = n
	}

	var total int32 = 0
	for c := head; c != nil; c = c.next {
		total += c.val
	}

	fmt.Printf("arena total = %d\n", total)
}

File: bench/zig/arena.zig
const std = @import("std");

const Node = struct {
    val: i32,
    next: ?*Node,
};

pub fn main() !void {
    const N: i32 = 5_000_000;

    var arena = std.heap.ArenaAllocator.init(std.heap.c_allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    var head: ?*Node = null;
    var i: i32 = 0;
    while (i < N) : (i += 1) {
        const n = try allocator.create(Node);
        n.* = .{ .val = i, .next = head };
        head = n;
    }

    var total: i32 = 0;
    var cur = head;
    while (cur) |c| {
        total +%= c.val;
        cur = c.next;
    }

    std.debug.print("arena total = {d}\n", .{total});
}
